crates/kitsune_p2p/kitsune_p2p/tests/sharding.rs

```diff
@@ -0,0 +1,387 @@
+
+    let tuner = |mut params: tuning_params_struct::KitsuneP2pTuningParams| {
+        params.gossip_arc_clamping = "none".to_string();
+        params.gossip_dynamic_arcs = false; // Don't update the arcs dynamically, use the initial value
+        params.disable_recent_gossip = true;
+        params.disable_historical_gossip = true;
+        params
+    };
+
+    // Arcs are this long by default, with an adjustment to ensure overlap.
+    let base_len = u32::MAX / 5;
+
+    let dim = SpaceDimension::standard();
+
+    let mut agents = Vec::new();
+
+    for i in 0..5 {
+        let mut harness = KitsuneTestHarness::try_new("")
+            .await
+            .expect("Failed to setup test harness")
+            .configure_tx5_network(signal_url)
+            .use_bootstrap_server(bootstrap_addr)
+            .update_tuning_params(tuner);
+
+        let sender = harness.spawn().await.expect("should be able to spawn node");
+
+        let mut agent = harness.create_agent().await;
+        let mut found_loc = false;
+        for _ in 0..1000 {
+            let loc = agent.get_loc().as_();
+            if loc > base_len * i && loc < base_len * (i + 1) {
+                found_loc = true;
+                break;
+            }
+
+            // If we didn't find a location in the right range, try again
+            agent = harness.create_agent().await;
+        }
+
+        assert!(
+            found_loc,
+            "Failed to find a location in the right range after 1000 tries"
+        );
+
+        // Distance to the end of the segment, plus the length of the next segment. Guaranteed to
+        // overlap with the next agent and not the one after that.
+        // Because of arc quantisation, the layout won't be perfect, but we can expect overlap at
+        // the start of the agent's arc, with the previous agent.
+        let len =
+            DhtLocation::new(base_len * (i + 1)) - agent.get_loc() + DhtLocation::new(base_len);
+
+        let arc = Arq::from_start_and_half_len_approximate(
+            dim,
+            &ArqStrat::standard(LocalStorageConfig::default(), 2.0),
+            agent.get_loc(),
+            len.as_() / 2 + 1,
+        );
+        sender
+            .join(space.clone(), agent.clone(), None, Some(arc))
+            .await
+            .unwrap();
+
+        agents.push((harness, sender, agent));
+    }
+
+    // Each agent should be connected to the next agent because that's how the arcs were set up
+    // above.
+    for i in 0..5 {
+        // A circular `next` so that the last agent is connected to the first agent
+        let next = (i + 1) % 5;
+
+        wait_for_connected(agents[i].1.clone(), agents[next].2.clone(), space.clone()).await
+    }
+
+    let sender_idx = 3;
+    let should_recv_idx = 2;
+
+    let sender_location = &agents[sender_idx].2 .0[32..];
+
+    let mut kitsune_basis = KitsuneBasis::new(vec![0; 36]);
+    kitsune_basis.0[32..].copy_from_slice(&sender_location);
```

Comment:
Someday we may add validation that the location is a correct hash of the hash... but I suppose we can address this somehow at that point (perhaps by having a test flag to disable that validation).